sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/InflatableX509ExtendedTrustManager.java

```diff
@@ -39,20 +52,95 @@ public class InflatableX509ExtendedTrustManager extends HotSwappableX509Extended
         try {
-            trustStore = KeyStoreUtils.createKeyStore();
+            this.trustStorePath = trustStorePath;
+            this.trustStorePassword = trustStorePassword;
+            this.trustStoreType = trustStoreType;
+
+            this.certificateAndAuthTypeTrustPredicate = Optional.ofNullable(certificateAndAuthTypeTrustPredicate)
+                    .orElse((chain, authType) -> false);
+
+            if (trustStorePath != null && trustStorePassword != null && StringUtils.isNotBlank(trustStoreType)) {
+                if (Files.exists(trustStorePath)) {
+                    trustStore = KeyStoreUtils.loadKeyStore(trustStorePath, trustStorePassword, trustStoreType);
+                } else {
+                    trustStore = KeyStoreUtils.createKeyStore(trustStoreType, trustStorePassword);
+                }
+            } else {
+                trustStore = KeyStoreUtils.createKeyStore();
+            }
         } finally {
             writeLock.unlock();
         }
     }
 
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        checkTrusted(trustManager -> super.checkServerTrusted(chain, authType), chain, authType);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        checkTrusted(trustManager -> super.checkServerTrusted(chain, authType, socket), chain, authType);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine) throws CertificateException {
+        checkTrusted(trustManager -> super.checkServerTrusted(chain, authType, sslEngine), chain, authType);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        checkTrusted(trustManager -> super.checkClientTrusted(chain, authType), chain, authType);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        checkTrusted(trustManager -> super.checkClientTrusted(chain, authType, socket), chain, authType);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine) throws CertificateException {
+        checkTrusted(trustManager -> super.checkClientTrusted(chain, authType, sslEngine), chain, authType);
+    }
+
+    private void checkTrusted(TrustManagerConsumer trustManagerConsumer, X509Certificate[] chain, String authType) throws CertificateException {
+        readLock.lock();
+
+        try {
+            trustManagerConsumer.checkTrusted(this);
+        } catch (CertificateException e) {
+            boolean shouldBeTrusted = certificateAndAuthTypeTrustPredicate.test(chain, authType);
+            if (shouldBeTrusted) {
+                addCertificates(Collections.singletonList(chain[0]));
```

Comment:
And finally the `addCertificates` will probably be in a dead lock state, since I think the current thread can be upgraded from read to write lock, but since there are many threads owning the read lock, I don't think it should work.

Reply:
I am not quite sure about that. I think it is still ok to have a readlock within the scope of this method as it is no issue to for all threads to call `trustManagerConsumer.checkTrusted(this);` However when there is a Certificate exception and it enters the catch block other threads needs to wait and that is now happening thanks to the synchronized block. Within that block it will enter the addCertificates method which contains a write lock. That method can also be accessed outside of this class so anyone needs to wait till that execution is finished if it is blocked. So my feeling is that there is no dead lock, but I would like to get your insight after testing it with the latest changes